### ISSUE TITLE
[DUBTE] Assignments i BoW

### ISSUE THREAD

**Issue by cmanso** [gdsa-upc/gdsa-suport#24]
Crec que hem fet bé totes les funcions i que ho fem correctament, pero cada cop que l'executem tenim diferents valors dels assignments i per tant del BoW, és normal? perquè em sembla molt extrany

Exemple:

[ 6  7  0  1  2  0  1  1  1  2  2  4  1  4  1  0  1  1  0  4  5  1  0  0  3
  1  4  3  1  0  1  1  1  0  0  0  2  0  0  0  2  0  0  5  3  5  1  0  1  1
  0  1  1  4  2  1 10  2  3  0  1  9  6  9  1  3  0  2  0  0  0  1  1  1  7
  4  0  1  1  3  1  0  1  2  6  0  0  2  7  0  0  2  1  5  3  0  1  5  5  5]
len: 100
dic_val: 50

a la següent execució, i per la mateixa imatge:

[ 9  1  3  3  5  2  3  2  1  1  1  1  0  1  1  3  1  4  0  1  7  1  1  0  0
  3  1  5  0  2  0  3  0  4  1  2  1  1  3  0  1  5  0  0  1  1  1  2  2  6
  0  1  1  1  2  2  5  1  1  0  2  1  3  0  1  8  1  2  0 10  1  0  0  2  1
  2  2  3  4  0  0  1  0  2  5  0  0  4  3  1  0  7  3  5 11  1  2  0  0  1]
len: 100
dic_val: 50

també la longuitud dels assignments i del bow em varia una mica, la hem posat a 100, però de vegades, per algunes imatges ens surt 99

**Comment by amaiasalvador** [comment]
Hola Carlos,

Possibles coses que poden estar passant i que podeu mirar:

1) Si cada cop que executeu el codi re-entreneu el codebook, pot ser que els centroides que obtingueu siguin diferents, ja que la inicialització és aleatòria, així que pot ser que els assignments canviin.

2) Mirant una mica el codi veig que utilitzeu la funció scipy.cluster.vq.whiten abans d'entrenar el codebook. Això és a vegades recomanable però no és necessari. La funció whiten transforma els teus descriptors dividint per la desviació standard del conjunt. Si després apliques aquesta mateixa funció als descriptors per separat, la transformació no serà la mateixa que abans, ja que la desviació standard serà diferent. Això no hauria de ser un problema, però proveu de saltar-vos aquest pas també.

3) Quant al tema de les longituds: el número d'assignments no té perquè ser 100. Ha de ser el número de descriptors locals que tingui una imatge. Reviseu aquesta part. El que ha de ser de tamany 100 és el vector bag of words. Penseu però que a l'hora de acumular els assigments per formar el vector, podria passar que hi hagi centroides sense representació. Igualment, el vector bow ha de ser de tamany 100, i les posicions corresponents als centroides sense representació han de ser 0. Veig que feu servir la funció numpy.bincount, que compta les repeticions dins d'una llista. Per això, quan li passeu una llista d'assigments on no tots les ids dels centroides apareixen us surten descriptors més curts.

**Comment by cmanso** [comment]
Moltes gràcies per la resposta, encara que ara tinc un altre dubte. Al build_bow, fent servir la funció preprocessing.normalize surten molts warnings:

/home/chiec/Enthought/Canopy_64bit/User/lib/python2.7/site-packages/sklearn/utils/validation.py:420: DataConversionWarning: Data with input dtype int64 was converted to float64 by the normalize function.
  warnings.warn(msg, DataConversionWarning)
/home/chiec/Enthought/Canopy_64bit/User/lib/python2.7/site-packages/sklearn/utils/validation.py:386: DeprecationWarning: Passing 1d arrays as data is deprecated in 0.17 and willraise ValueError in 0.19. Reshape your data either using X.reshape(-1, 1) if your data has a single feature or X.reshape(1, -1) if it contains a single sample.
  DeprecationWarning)

no sé si és que ho estic fent malament o que hauría de normalitzar tots els BoW de cop.

Gràcies.

**Comment by amaiasalvador** [comment]
Hola,

Ja ho fas bé, els vectors BoW s'han de normalitzar un per un. Segurament aquest warning et surt perquè el teu descriptor d'entrada a la funció normalize té només una dimensió. Si mires la documentació de la funció [normalize](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.normalize.html), aquesta espera que li passis un conjunt de descriptors de tamany [n_samples, n_features]. En el teu cas, com que normalitzes els BoW per separat, hauràs de tenir vectors de [1,100], on 100 és el tamany del codebook. Segurament el que et passa és que en comptes de [1,100] tens vectors de [100], i per això es queixa. Ho pots arreglar fàcilment amb [numpy.reshape](http://docs.scipy.org/doc/numpy-1.10.1/reference/generated/numpy.reshape.html).
